### PR TITLE
Generate an "UpdateXXX" method for Entity, similar to Scala's copy functionality for its case classes.

### PR DESCRIPTION
**Pull request by shatterboom**
### tl;dr

I implemented the idea presented here #111

A new UpdateXXX method is generated. It performs the same role as ReplaceXX but it uses C# default parameters and nullables.
### Nicer syntax

It allows a nicer syntax than ReplaceXXX.

```
TestComponent:IComponent{
     public string name;
     public int height;
}

public Entity ReplaceTest(string newName, int newHeight){
    var component = CreateComponent<TestComponent>(ComponentIds.Test);
    component.height = newHeight;
    component.name= newName;
    ReplaceComponent(ComponentIds.Test, component);
    return this;
 }

public Entity UpdateTest(string name = null,int? height = null){
    var component = CreateComponent<TestComponent>(ComponentIds.Test);
    component.height  = height .GetValueOrDefault(component.height );
    component.name= name;
    ReplaceComponent(ComponentIds.Test, component);
    return this;
}
```

Usually, if we want to modify`TestComponent.height` we need to do: `e.ReplaceTest(e.test.name,42);`
With the UpdateXXX method, it can be simplified to: `e.UpdateTest(height:42);`
### Nicer refactoring

It also provides a couple of advantages regarding refactoring:
- The order of the parameters doesn't necessarily depend on the order of the fields in the component. It avoids sneaky bugs when fields of the same type are reordered.
- Adding new fields to a components doesn't require to modify all the calls to ReplaceXXX. 
### Performances

I did a [basic benchmark](https://gist.github.com/shatterboom/8dfff936a119d151516da6a590caa493) and I found that the performances are close. The only difference with ReplaceXXX is the extra calls to `GetValueOrDefault`, like:

`component.x = newX.GetValueOrDefault(component.x)`
instead of
`component.x = newX;`

Therefore, the impact increases with the number of nullable fields. For non value type, UpdateXXX will not use GetValueOrDefault.

I didn't find any extra garbage (explanation [here](https://stackoverflow.com/questions/7131910/c-sharp-nullable-value-type-generates-garbage))
### AddXXX

This idea could be applied to the AddXXX method as well but the correct initialization of components wouldn't be guaranteed.

### CODE REVIEW

**Comment by gjroelofs** [comment]
Just a heads up; the Unity Mono-based compiler has definite issues when it comes to the ordering of named arguments that leads to the compiler itself crashing.

https://gist.github.com/gjroelofs/626a67f22a6956e361c9

**Comment by shatterboom** [comment]
@gjroelofs Thanks for the heads up. Fortunately, @sschmid said that is main focus is the Roselyn Compiler :)

**Comment by gjroelofs** [comment]
It's not about the compiler used to analyze a project though; this is about the compiler used by Unity itself to compile your project. I suspect that @sschmid isn't supporting replacing the compiler in Unity as this isn't popularly supported.  (Although we ourselves use: https://github.com/SaladLab/Unity3D.IncrementalCompiler)

**Comment by sschmid** [comment]
Hi, long time no see.
It's really easy now to add you own custom code generators (see #265). This means you can simply duplicate the a current generator (e.g. [ComponentEntityGenerator.cs](https://github.com/sschmid/Entitas-CSharp/blob/develop/Plugins/Entitas.CodeGenerator/Entitas.CodeGenerator/Generators/ComponentEntityGenerator.cs)) and change it to produce the code you want to have. Once you're done, you can select your code generator in the Entitas Preference Window in Unity.
The default generator won't be updated for now, but feel free to create your own one and create a new pull request if you think that should be part of Entitas.
Thanks!
